### DISCUSSION TITLE
Is it possible to add custom transformations for files before compilation?

### DISCUSSION THREAD

**Discussion by mainrs** [tectonic-typesetting/tectonic#968] [Q&A]
Similar to `latexmk`. I have some `.txt` files that I want to convert to latex before compiling it with `tectonic`.

**Comment by Neved4** [comment]
Hey @mainrs! How do you use `latexmk` to convert `.txt` files? Could you explain me a bit more about your use case?

**Reply by harrymander** [reply]
Not sure of @mainrs's exact use case, but I use `latexmk` similarly for generating graphics from scripts, e.g. converting SVG to PDF with Inkscape or generating PGF plots with `pyplot`. I have just recently discovered this project and am certainly interested in using it since it seems much simpler than `latexmk`!

Here is a simple example using pyplot:

`example.txt`
```tex
\documentclass[a4paper]{article}

\usepackage{pgfplots}

\begin{document}

Hello, World

\input{plot.pgf}

\end{document}
```

`plot.py`
```python
import sys
import matplotlib.pyplot as plt


if __name__ == '__main__':
    path = sys.argv[1]
    plt.figure()
    plt.plot(range(50))
    plt.savefig(path, format='pgf')
```

`latexmkrc` (perl)
```perl
@default_files = ('example.tex');
$pdf_mode = 1;
$pdflatex = 'pdflatex --shell-escape -interaction=batchmode -synctex=1 %O %S';
$cleanup_includes_generated = 1;
$cleanup_includes_cusdep_generated = 1;

add_cus_dep("py", "pgf", 0, "py2pgf");
sub py2pgf {
    system("python3 $_[0].py $_[0].pgf");
}
```

`latexmk` will attempt to compile `example.tex` with `pdflatex` and get an error saying that  `plot.pgf` is missing. It will then look for  a corresponding `plot.py` to generate the PGF file before running `pdflatex` again.

Alternatively, you could generate the dependencies with `make` and incorporate `latexmk` into it. I suppose you could also incorporate `tectonic` into a `Makefile` as well.
